### ISSUE TITLE
Html Agility Pack performance on .NET Core

### ISSUE THREAD

**Issue by raiytu4** [zzzprojects/html-agility-pack#207]
I have recently moved my project from .net farmwork to .net core.
Although .NET Core 's performance is way better that .NET Framework,
Html Agility Pack on .NET Core turn out slower than on .NET Framework.
( you can read about .net core performance vs .net farmwork here https://blogs.msdn.microsoft.com/dotnet/2017/06/07/performance-improvements-in-net-core/)
Specifically when I run this code on a simple console app:
```
class Program
    {
        static void Main(string[] args)
        {
            HtmlWeb web = new HtmlWeb();

            var doc = web.Load("http://sstruyen.com/doc-truyen/he-thong-hac-khoa-ky-quan-net/15375.html").DocumentNode;
            var sw = new Stopwatch();
            sw.Start();
            for (int i = 0; i < 10000; ++i)
            {
                FetchStoryTitle(doc);
            }
            sw.Stop();
            Console.WriteLine(sw.Elapsed.TotalSeconds);
            Console.ReadLine();
        }

        public static string FetchStoryTitle(HtmlNode doc)
        {
            string xpath = "//*[@id=\"wrapper\"]/div/div/div/div" +
                "/div[contains(@class, 'fixContent')]" +
                "/div[contains(@class, 'truyeninfo')]" +
                "/h1[contains(@class, 'title')]/a";
            var titleNode = doc.SelectSingleNode(xpath);
            if (titleNode != null)
            {
                var text = titleNode.InnerText.Trim();
                if (text != null && text.Any())
                {
                    return text;
                }
            }
            return null;
        }
    }
```
It's take 6,9355134s on .net farmwork and 56,8355829 on .net core. That almost 10 time slower!
Any idea why this happens?

**Comment by JonathanMagnan** [comment]
Hello @raiytu4 ,

Are you sure you are testing the same code/amount of iterations?

On debug mode, it takes on my side 40s for both of them.

Best Regards,

Jonathan

---

**Performance Libraries**
`context.BulkInsert(list, options => options.BatchSize = 1000);  `
[Entity Framework Extensions](http://entityframework-extensions.net/) • [Bulk Operations](http://bulk-operations.net/) •  [Dapper Plus](http://dapper-plus.net/) •  [LinqToSql Plus](http://linqtosql-plus.net/)

**Runtime Evaluation**
`Eval.Execute("x + y", new {x = 1, y = 2}); // return 3`
[C# Eval Function](http://eval-expression.net/) • [SQL Eval Function](http://eval-sql.net/)

**Comment by raiytu4** [comment]
Hello @JonathanMagnan ,

I double checked everything,  same code, same amount of iterations ( I literally Ctrl + C, Ctrl + V you can check you my source code here https://github.com/raiytu4/SpeedTestCoreVsDotnet)
I turn off everything and run the code one by one in cmd, this time I trying out on the release build, slightly faster but same problem
![capture](https://user-images.githubusercontent.com/25586304/41384599-fdded7ec-6fa0-11e8-9a5f-f2692a86e558.PNG)

Best Regards,

Kiet

**Comment by raiytu4** [comment]
Hello @JonathanMagnan 

Thank for the quick response, I think I found out the solution for this.

At first, when I discover this problem I also anticipate the problem is related to XPath XML stuff since your core library is depend on XML and XPath stuff.

The solution for this is switching to CSS selector instead. As you mention in #141 I tried out those library and do a quick performance test. Turn out CSS selector is way faster than XPath. With the same test as above, CSS selector gets only 1.8s in .net core and 2.1s in .net framework and in debug mode ( much be even faster when running separately with release build). 

In fact, I had the intention to open a issues asking you to support CSS selector in the past. Not to mention the speed, as my experience CSS selector is more convenient than XPath in many away ( I may describe more if you interacting about it) 

Overall, I still think this a great library, very enthusiastic support. It's helped me a lot. Thank you!

Best Regards,

Kiet
